Add Route53 health checks

jsf9k opened the pull request:
## 🗣 Description ##

This pull request modifies the code to utilize Route53 health checks to automatically update DNS records when FreeIPA server instances become unhealthy.

## 💭 Motivation and context ##

In addition to being a good idea, this improvement may help with cisagov/cool-system-internal#89.  If it does not then I have some ideas for similar improvements based on custom CloudWatch metrics that will definitely help.

## 🧪 Testing ##

All automated testing passes.  I have deployed these changes to the SharedServices account in the staging COOL environment.

## ✅ Pre-approval checklist ##

- [x] This PR has an informative and human-readable title.
- [x] Changes are limited to a single goal - *eschew scope creep!*
- [x] All relevant type-of-change labels have been added.
- [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
- [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
- [x] All relevant repo and/or project documentation has been updated to reflect the changes in this PR.
- [x] All new and existing tests pass.

## ✅ Pre-merge checklist ##

- [x] Revert dependencies to default branches.

jsf9k committed:
Rename a Terraform file

The new name better describes what kind of resources the file
contains.

jsf9k committed:
Add CloudWatch alarms for the FreeIPA instances

jsf9k committed:
Alphabetize module meta-arguments

jsf9k committed:
Move DNS records out of module

These records will soon need to be tied more tightly to the CW alarms.

jsf9k committed:
Alphabetize resource parameters

jsf9k committed:
Add Route53 health checks and tie them into DNS records

jsf9k committed:
Combine two health checks into a calculated overall health check

The two health checks being combined are the system and instance
status checks.

jsf9k committed:
Remove unused input variable

jsf9k committed:
Update README.md via terraform-docs

jsf9k committed:
Add @mcdonnnj as a codeowner

dav3r reviewed (approved):
Super sweet!  💪 💼

mcdonnnj commented on dns_records.tf:
```diff
@@ -0,0 +1,316 @@
+  reverse_zone_ids = [for zone in data.terraform_remote_state.networking.outputs.private_subnet_private_reverse_zones : zone.id]
+}
+
+resource "aws_route53_record" "server0_A" {
```
Are we specifically avoiding doing one `aws_route53_record` with the `count` meta-argument for each type of record? Since `count.index` starts at `0` and that aligns with record names I'm just curious if there's a reason/rationale for avoiding consolidation.

jsf9k replied:
I don't like to use `count` because it sometimes leads to an ordering problem, where resources get recreated just because they are now `1` and they used to be `0`.  I prefer to use a `map` with keys that tie to the actual resource.

Please see commits d2eafb6 and f32a936.

mcdonnnj replied:
That's fair. It just scratched at my DRY urges. I appreciate the consolidation!

mcdonnnj commented on README.md:
```diff
@@ -52,7 +52,7 @@ module "example" {
 
```
I was going to mention updating the `Notes` section for the removal of the `dns` server when I noticed that neither `dns` nor the `security_groups` folders are mentioned. Would you update it to reflect the `security_groups` directory?

jsf9k replied:
Are you referring to the need to run `terraform init` on subdirectories?  This is not necessary for _in situ_ submodules; you can run `pre-commit run -a` locally to verify this.

mcdonnnj replied:
Ahh right, because it gets treated as part of the configuration in the root directory through usage. Apologies.

mcdonnnj commented on dns_health_checks.tf:
```diff
@@ -0,0 +1,53 @@
+
+resource "aws_route53_health_check" "system_status_check" {
+  for_each = {
+    "${module.ipa0.server.id}" = { hostname = "ipa0", server = module.ipa0.server }
```
Is the `server` key for future use? (This question applies to all of these resources)

jsf9k replied:
I thought I would require that attribute, but it ended up that I did not.  Please see commit a695740.

mcdonnnj reviewed (approved):
This will be quite helpful. I did have a couple of questions and a request to update the README with some missing information.

jsf9k committed:
Remove unnecessary attribute from for_each map

I thought I would need the attribute, but as it turned out I did not.

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

jsf9k committed:
Combine multiple records via for_each

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

jsf9k committed:
Remove unnecessary attribute from for_each maps

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

jsf9k committed:
Use the default branch of cisagov/instance-cw-alarms-tf-module

The branch we were using was merged into the default branch via
cisagov/instance-cw-alarms-tf-module#5.
